core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1919,18 +1919,24 @@ class SparkContext(config: SparkConf) extends Logging {
           // A JAR file which exists only on the driver node
           case "file" => addLocalJarFile(new File(uri.getPath))
           // A JAR file which exists locally on every worker node
-          case "local" => "file:" + uri.getPath
+          case "local" => Array("file:" + uri.getPath)
+          case "ivy" =>
+            // Since `new Path(path).toUri` will lose query information,
+            // so here we use `URI.create(path)`
+            DependencyUtils.resolveMavenDependencies(URI.create(path)).split(",")
           case _ => checkRemoteJarFile(path)
         }
       }
-      if (key != null) {
+      if (keys.nonEmpty) {
         val timestamp = if (addedOnSubmit) startTime else System.currentTimeMillis
-        if (addedJars.putIfAbsent(key, timestamp).isEmpty) {
-          logInfo(s"Added JAR $path at $key with timestamp $timestamp")
-          postEnvironmentUpdate()
-        } else {
-          logWarning(s"The jar $path has been added already. Overwriting of added jars " +
-            "is not supported in the current version.")
+        keys.foreach { key =>
+          if (addedJars.putIfAbsent(key, timestamp).isEmpty) {
+            logInfo(s"Added JAR $path at $key with timestamp $timestamp")
+            postEnvironmentUpdate()
+          } else {
+            logWarning(s"The jar $path has been added already. Overwriting of added jars " +
```

Comment:
If `transitive=true` and there are too many dependencies, users possibly see a lot of warning messages?

Reply:
> If `transitive=true` and there are too many dependencies, users possibly see a lot of warning messages?

Also `logInfo(s"Added JAR $path at $key with timestamp $timestamp")`

How about like 
```
val (added, existed) = keys.partition(addedJars.putIfAbsent(_, timestamp).isEmpty)
        if (added.nonEmpty) {
          logInfo(s"Added JAR $path at ${added.mkString(",")} with timestamp $timestamp")
          postEnvironmentUpdate()
        }
        if(existed.nonEmpty) {
          logWarning(s"The dependencies jars ${existed.mkString(",")} of $path has been added" +
            s" already. Overwriting of added jars is not supported in the current version.")
        }
```

Reply:
Looks fine now, but I think we need to make the error message more general because `dependencies` is only meaningful for the `ivy` schema case.

Reply:
> Looks fine now, but I think we need to make the error message more general because `dependencies` in the message is only meaningful in the `ivy` schema case.

Updated